include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -75,18 +75,19 @@ struct maybe_remove_property<PoolResource,
                              Upstream,
                              Property,
                              cuda::std::enable_if_t<!cuda::has_property<Upstream, Property>>> {
-#ifdef __GNUC__  // GCC warns about compatibility issues with pre ISO C++ code
+#if defined(__GNUC__) && !defined(__clang__)  // GCC warns about compatibility
```

Comment:
Just thought of this: Under what circumstances are `__GNUC__` and `__clang__` both defined? They must be or we wouldn't need this logic?

Reply:
- `__GNUC__` means "compatibility with GNU C". This is defined by both GCC and Clang.
- `__clang__` means Clang. I'd guess that this is also defined by clang derivatives such as HIPCC and DPC++. But it's not defined by GCC.
- AFAIK GCC doesn't have a similar macro. I.e. there is no `__gcc__` macro.

> Under what circumstances are `__GNUC__` and `__clang__` both defined?

When you run a compilation with clang this is the case.

The `"-Wnon-template-friend"` exists in GCC, but not in Clang. With the way things are now we're essentially saying: If you're a GNU C compatible compiler, but you're not Clang (or a Clang derivative where we can expect warning flags to be reasonably similar to clang), then add `-Wnon-template-friend` to the ignorelist.

Before this patch Clang triggered warnings like `warning: ignoring unknown warning` which spammed the logs.

Reply:
Thanks for the informative answer!